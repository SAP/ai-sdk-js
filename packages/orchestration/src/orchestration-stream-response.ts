import type { ToolCallAccumulator } from './util/index.js';
import type {
  MessageToolCalls,
  TokenUsage
} from './client/api/schema/index.js';
import type { OrchestrationStream } from './orchestration-stream.js';

/**
 * Orchestration stream response.
 */
export class OrchestrationStreamResponse<T> {
  private _usage: TokenUsage | undefined;
  /**
   * Finish reasons for all choices.
   */
  private _finishReasons: Map<number, string> = new Map();
  private _toolCallsAccumulators: Map<
    number,
    Map<number, ToolCallAccumulator>
  > = new Map();
  private _stream: OrchestrationStream<T> | undefined;
  private _toolCalls: Map<number, MessageToolCalls> = new Map();

  /**
   * Gets the token usage for the response.
   * @returns The token usage for the response.
   */
  public getTokenUsage(): TokenUsage | undefined {
    return this._usage;
  }

  /**
   * @internal
   */
  _setTokenUsage(usage: TokenUsage): void {
    this._usage = usage;
  }

  /**
   * Gets the finish reason for a specific choice index.
   * @param choiceIndex - The index of the choice to get the finish reason for.
   * @returns The finish reason for the specified choice index.
   */
  public getFinishReason(choiceIndex = 0): string | undefined {
    return this._finishReasons.get(choiceIndex);
  }

  /**
   * @internal
   */
  _getFinishReasons(): Map<number, string> {
    return this._finishReasons;
  }

  /**
   * @internal
   */
  _setFinishReasons(finishReasons: Map<number, string>): void {
    this._finishReasons = finishReasons;
  }

  /**
   * Gets the tool calls for a specific choice index.
   * @param choiceIndex - The index of the choice to get the tool calls for.
   * @returns The tool calls for the specified choice index.
   */
  public getToolCalls(choiceIndex = 0): MessageToolCalls | undefined {
    return this._toolCalls.get(choiceIndex);
  }

  /**
   * @internal
   */
  _setToolCalls(choiceIndex: number, toolCalls: MessageToolCalls): void {
    this._toolCalls.set(choiceIndex, toolCalls);
  }

  /**
   * @internal
   */
  _getToolCallsAccumulators(): Map<number, Map<number, ToolCallAccumulator>> {
    return this._toolCallsAccumulators;
  }

  get stream(): OrchestrationStream<T> {
    if (!this._stream) {
      throw new Error('Response stream is undefined.');
    }
    return this._stream;
  }

  /**
   * @internal
   */
  set stream(stream: OrchestrationStream<T>) {
    this._stream = stream;
  }
}
