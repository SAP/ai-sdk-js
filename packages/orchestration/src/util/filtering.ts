import { supportedAzureFilterThresholds } from '../orchestration-types.js';
import type {
  AzureContentSafety,
  AzureContentSafetyFilterConfig,
  InputFilteringConfig,
  LlamaGuard38BFilterConfig,
  OutputFilteringConfig
} from '../client/api/schema/index.js';
import type {
  AzureContentFilter,
  AzureFilterThreshold,
  LlamaGuardCategory
} from '../orchestration-types.js';

/**
 * Convenience function to build Azure content filter.
 * @param filter - Filtering configuration for Azure filter. If skipped, the default Azure content filter configuration is used.
 * @returns An object with the Azure filtering configuration.
 * @deprecated Since 1.8.0. Use {@link buildAzureContentSafetyFilter()} instead.
 */
export function buildAzureContentFilter(
  filter?: AzureContentSafety
): InputFilteringConfig | OutputFilteringConfig {
  if (filter && !Object.keys(filter).length) {
    throw new Error('Filter property cannot be an empty object');
  }
  return {
    filters: [
      {
        type: 'azure_content_safety',
        ...(filter && { config: filter })
      }
    ]
  };
}

/**
 * Convenience function to build Azure content filter.
 * @param config - Configuration for Azure content safety filter.
 * If skipped, the default configuration of `ALLOW_SAFE_LOW` is used for all filter categories.
 * @returns Filter config object.
 * @example "buildAzureContentSafetyFilter({ Hate: 'ALLOW_SAFE', Violence: 'ALLOW_SAFE_LOW_MEDIUM' })"
 */
export function buildAzureContentSafetyFilter(
  config?: AzureContentFilter
): AzureContentSafetyFilterConfig {
  if (config && !Object.keys(config).length) {
    throw new Error('Filtering configuration cannot be an empty object');
  }
  return {
    type: 'azure_content_safety',
    ...(config && {
      config: {
        ...Object.fromEntries(
          Object.entries(config).map(([key, value]) => [
            key,
            supportedAzureFilterThresholds[value as AzureFilterThreshold]
          ])
        )
      }
    })
  };
}

/**
 * Convenience function to build Llama guard filter.
 * @param categories - Categories to be enabled for filtering. Provide at least one category.
 * @returns Filter config object.
 * @example "buildLlamaGuardFilter('self_harm', 'hate')"
 */
export function buildLlamaGuardFilter(
  ...categories: [LlamaGuardCategory, ...LlamaGuardCategory[]]
): LlamaGuard38BFilterConfig {
  return {
    type: 'llama_guard_3_8b',
    config: Object.fromEntries(
      [...categories].map(category => [category, true])
    )
  };
}
