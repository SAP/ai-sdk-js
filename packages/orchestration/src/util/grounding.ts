import type { GroundingModuleConfig } from '../client/api/schema/index.js';
import type { DocumentGroundingServiceConfig } from '../orchestration-types.js';

/**
 * Convenience function to create Document Grounding configuration.
 * @param groundingConfig - Configuration for the document grounding service.
 * @returns An object with the full grounding configuration.
 */
export function buildDocumentGroundingConfig(
  groundingConfig: DocumentGroundingServiceConfig
): GroundingModuleConfig {
  return {
    type: 'document_grounding_service',
    config: {
      input_params: groundingConfig.input_params,
      output_param: groundingConfig.output_param,
      ...(groundingConfig.filters && {
        filters: groundingConfig.filters?.map(filter => ({
          data_repository_type: 'vector',
          ...filter
        }))
      })
    }
  };
}
