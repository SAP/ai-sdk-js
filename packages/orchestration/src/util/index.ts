export * from './filtering.js';
export * from './grounding.js';
export * from './module-config.js';
export * from './masking.js';
export * from './translation.js';
export * from './tool-calls.js';
