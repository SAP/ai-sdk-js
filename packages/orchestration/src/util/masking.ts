import type { DpiConfig } from '../client/api/schema/index.js';
import type { DpiMaskingConfig } from '../orchestration-types.js';

/**
 * Convenience function to create masking provider SAP Data Privacy Integration.
 * @param dpiMaskingConfig - Configuration for the masking provider SAP Data Privacy Integration.
 * @returns An object with the masking provider configuration.
 */
export function buildDpiMaskingProvider(
  dpiMaskingConfig: DpiMaskingConfig
): DpiConfig {
  const { method, mask_grounding_input, entities, allowlist } =
    dpiMaskingConfig;
  return {
    type: 'sap_data_privacy_integration',
    method,
    entities: entities.map(entity => {
      if (typeof entity === 'string') {
        return { type: entity };
      }
      return entity.type === 'custom'
        ? (({ type: _, ...rest }) => rest)(entity)
        : entity;
    }),
    ...(mask_grounding_input !== undefined && {
      mask_grounding_input: {
        enabled: mask_grounding_input
      }
    }),
    ...(allowlist && {
      allowlist
    })
  };
}
