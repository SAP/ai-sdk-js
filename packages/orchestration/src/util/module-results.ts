/**
 * @internal
 */
export function mergeModuleResults<T>(
  ...results: (T | undefined)[]
): T | undefined {
  return results.reduce((acc, result) => {
    if (result === undefined) {
      return acc;
    }
    if (acc === undefined) {
      return result;
    }
    return { ...acc, ...result };
  }, undefined as T | undefined);
}
