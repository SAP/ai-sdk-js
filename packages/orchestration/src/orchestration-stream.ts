import { createLogger } from '@sap-cloud-sdk/util';
import { SseStream } from '@sap-ai-sdk/core';
import { OrchestrationStreamChunkResponse } from './orchestration-stream-chunk-response.js';
import type {
  CompletionPostResponseStreaming,
  LlmChoiceStreaming
} from './client/api/schema/index.js';
import type { HttpResponse } from '@sap-cloud-sdk/http-client';
import type { OrchestrationStreamResponse } from './orchestration-stream-response.js';

const logger = createLogger({
  package: 'orchestration',
  messageContext: 'orchestration-chat-completion-stream'
});

/**
 * Orchestration stream containing post-processing functions.
 */
export class OrchestrationStream<Item> extends SseStream<Item> {
  /**
   * Create an orchestration stream based on the http response.
   * @param response - Http response.
   * @returns An orchestration stream.
   * @internal
   */
  public static _create(
    response: HttpResponse,
    controller: AbortController
  ): OrchestrationStream<CompletionPostResponseStreaming> {
    const stream =
      SseStream.transformToSseStream<CompletionPostResponseStreaming>(
        response,
        controller
      );
    return new OrchestrationStream(stream.iterator, controller);
  }

  /**
   * Wrap raw chunk data with chunk response class to provide helper functions.
   * @param stream - Orchestration stream.
   * @internal
   */
  static async *_processChunk(
    stream: OrchestrationStream<CompletionPostResponseStreaming>
  ): AsyncGenerator<OrchestrationStreamChunkResponse> {
    for await (const chunk of stream) {
      yield new OrchestrationStreamChunkResponse(chunk);
    }
  }

  /**
   * @internal
   */
  static async *_processFinishReason(
    stream: OrchestrationStream<OrchestrationStreamChunkResponse>,
    response?: OrchestrationStreamResponse<OrchestrationStreamChunkResponse>
  ): AsyncGenerator<OrchestrationStreamChunkResponse> {
    for await (const chunk of stream) {
      chunk.data.orchestration_result?.choices.forEach(
        (choice: LlmChoiceStreaming) => {
          const choiceIndex = choice.index;
          if (choiceIndex >= 0) {
            const finishReason = chunk.getFinishReason(choiceIndex);
            if (finishReason) {
              if (response) {
                response._getFinishReasons().set(choiceIndex, finishReason);
              }
              switch (finishReason) {
                case 'content_filter':
                  logger.error(
                    `Choice ${choiceIndex}: Stream finished with content filter hit.`
                  );
                  break;
                case 'length':
                  logger.error(
                    `Choice ${choiceIndex}: Stream finished with token length exceeded.`
                  );
                  break;
                case 'stop':
                  logger.debug(`Choice ${choiceIndex}: Stream finished.`);
                  break;
                case 'tool_calls':
                  logger.error(
                    `Choice ${choiceIndex}: Stream finished with tool calls exceeded.`
                  );
                  break;
                case 'function_call':
                  logger.error(
                    `Choice ${choiceIndex}: Stream finished with function call exceeded.`
                  );
                  break;
                default:
                  logger.error(
                    `Choice ${choiceIndex}: Stream finished with unknown reason '${finishReason}'.`
                  );
              }
            }
          }
        }
      );
      yield chunk;
    }
  }

  /**
   * @internal
   */
  static async *_processTokenUsage(
    stream: OrchestrationStream<OrchestrationStreamChunkResponse>,
    response?: OrchestrationStreamResponse<OrchestrationStreamChunkResponse>
  ): AsyncGenerator<OrchestrationStreamChunkResponse> {
    for await (const chunk of stream) {
      const usage = chunk.getTokenUsage();
      if (usage) {
        if (response) {
          response._setTokenUsage(usage);
        }
        logger.debug(`Token usage: ${JSON.stringify(usage)}`);
      }
      yield chunk;
    }
  }

  /**
   * Transform a stream of chunks into a stream of content strings.
   * @param stream - Orchestration stream.
   * @param choiceIndex - The index of the choice to parse.
   * @internal
   */
  static async *_processContentStream(
    stream: OrchestrationStream<OrchestrationStreamChunkResponse>
  ): AsyncGenerator<string> {
    for await (const chunk of stream) {
      const deltaContent = chunk.getDeltaContent();
      if (!deltaContent) {
        continue;
      }
      yield deltaContent;
    }
  }

  constructor(
    public iterator: () => AsyncIterator<Item>,
    controller: AbortController
  ) {
    super(iterator, controller);
  }

  /**
   * Pipe the stream through a processing function.
   * @param processFn - The function to process the input stream.
   * @param response - The `OrchestrationStreamResponse` object for process function to store finish reason, token usage, etc.
   * @returns The output stream containing processed items.
   * @internal
   */
  _pipe<TReturn>(
    processFn: (
      stream: OrchestrationStream<Item>,
      response?: OrchestrationStreamResponse<OrchestrationStreamChunkResponse>
    ) => AsyncIterator<TReturn>,
    response?: OrchestrationStreamResponse<OrchestrationStreamChunkResponse>
  ): OrchestrationStream<TReturn> {
    if (response) {
      return new OrchestrationStream(
        () => processFn(this, response),
        this.controller
      );
    }
    return new OrchestrationStream(() => processFn(this), this.controller);
  }

  /**
   * Transform the stream of chunks into a stream of content strings.
   * @param this - Orchestration stream.
   * @returns A stream of content strings.
   */
  public toContentStream(
    this: OrchestrationStream<OrchestrationStreamChunkResponse>
  ): OrchestrationStream<string> {
    return new OrchestrationStream(
      () => OrchestrationStream._processContentStream(this),
      this.controller
    );
  }
}
