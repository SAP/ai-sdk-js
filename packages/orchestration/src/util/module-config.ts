import { createLogger } from '@sap-cloud-sdk/util';
import type {
  Prompt,
  StreamOptions,
  LlmModuleConfig,
  OrchestrationModuleConfig
} from '../orchestration-types.js';
import type {
  CompletionPostRequest,
  FilteringStreamOptions,
  ModuleConfigs,
  OrchestrationConfig,
  OutputFilteringConfig,
  GlobalStreamOptions,
  TemplatingModuleConfig
} from '../client/api/schema/index.js';

const logger = createLogger({
  package: 'orchestration',
  messageContext: 'orchestration-utils'
});

/**
 * @internal
 */
export function constructCompletionPostRequestFromJsonModuleConfig(
  config: Record<string, any>,
  prompt?: Prompt,
  stream?: boolean
): Record<string, any> {
  const orchestration_config = { ...config };
  if (stream) {
    orchestration_config.stream = true;
  } else {
    delete orchestration_config.stream;
  }

  return {
    messages_history: prompt?.messagesHistory || [],
    input_params: prompt?.inputParams || {},
    orchestration_config
  };
}

/**
 * @internal
 */
export function addStreamOptionsToLlmModuleConfig(
  llmModuleConfig: LlmModuleConfig,
  streamOptions?: StreamOptions
): LlmModuleConfig {
  if (streamOptions?.llm === null) {
    return llmModuleConfig;
  }
  return {
    ...llmModuleConfig,
    model_params: {
      ...llmModuleConfig.model_params,
      ...(streamOptions?.llm !== null && {
        stream_options: {
          include_usage: true,
          ...(llmModuleConfig.model_params?.stream_options || {}),
          ...(streamOptions?.llm || {})
        }
      })
    }
  };
}

/**
 * @internal
 */
export function addStreamOptionsToOutputFilteringConfig(
  outputFilteringConfig: OutputFilteringConfig,
  filteringStreamOptions: FilteringStreamOptions
): OutputFilteringConfig {
  return {
    ...outputFilteringConfig,
    stream_options: {
      ...(outputFilteringConfig.stream_options || {}),
      ...filteringStreamOptions
    }
  };
}

/**
 * @internal
 */
export function addStreamOptions(
  moduleConfigs: ModuleConfigs,
  streamOptions?: StreamOptions
): OrchestrationConfig {
  const { llm_module_config, filtering_module_config } = moduleConfigs;
  const outputFiltering = streamOptions?.outputFiltering;
  const globalOptions = streamOptions?.global;

  if (!moduleConfigs?.filtering_module_config?.output && outputFiltering) {
    logger.warn(
      'Output filter stream options are not applied because filtering module is not configured.'
    );
  }

  return {
    stream: true,
    ...(globalOptions && { stream_options: globalOptions }),
    module_configurations: {
      ...moduleConfigs,
      llm_module_config: addStreamOptionsToLlmModuleConfig(
        llm_module_config,
        streamOptions
      ),
      ...(outputFiltering &&
        filtering_module_config?.output && {
          filtering_module_config: {
            ...filtering_module_config,
            output: addStreamOptionsToOutputFilteringConfig(
              filtering_module_config.output,
              outputFiltering
            )
          }
        })
    }
  };
}

/**
 * @internal
 */
export function constructCompletionPostRequest(
  config: OrchestrationModuleConfig,
  prompt?: Prompt,
  stream?: boolean,
  streamOptions?: StreamOptions
): CompletionPostRequest {
  const moduleConfigurations: ModuleConfigs = {
    templating_module_config: config.templating as TemplatingModuleConfig,
    llm_module_config: config.llm,
    ...(config?.filtering &&
      Object.keys(config.filtering).length && {
        filtering_module_config: config.filtering
      }),
    ...(config?.masking &&
      Object.keys(config.masking).length && {
        masking_module_config: config.masking
      }),
    ...(config?.grounding &&
      Object.keys(config.grounding).length && {
        grounding_module_config: config.grounding
      })
  };

  return {
    orchestration_config: stream
      ? addStreamOptions(
          moduleConfigurations,
          mergeStreamOptions(config.streaming, streamOptions)
        )
      : { module_configurations: moduleConfigurations },
    ...(prompt?.inputParams && {
      input_params: prompt.inputParams
    }),
    ...(prompt?.messagesHistory && {
      messages_history: prompt.messagesHistory
    })
  };
}

function mergeStreamOptions(
  globalOptions?: GlobalStreamOptions,
  streamOptions?: StreamOptions
): StreamOptions {
  return {
    ...streamOptions,
    ...((globalOptions || streamOptions?.global) && {
      global: {
        ...globalOptions,
        ...streamOptions?.global
      }
    })
  };
}
