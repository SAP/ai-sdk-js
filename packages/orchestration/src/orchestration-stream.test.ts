import { createLogger } from '@sap-cloud-sdk/util';
import { jest } from '@jest/globals';
import { LineDecoder, SSEDecoder } from '@sap-ai-sdk/core';
import { parseFileToString } from '../../../test-util/mock-http.js';
import { OrchestrationStream } from './orchestration-stream.js';
import { OrchestrationStreamResponse } from './orchestration-stream-response.js';
import type { CompletionPostResponseStreaming } from './client/api/schema/index.js';

describe('Orchestration chat completion stream', () => {
  let sseChunks: string[];
  let originalChatCompletionStream: OrchestrationStream<CompletionPostResponseStreaming>;

  beforeEach(async () => {
    const rawChunksString = await parseFileToString(
      'orchestration',
      'orchestration-chat-completion-stream-chunks.txt'
    );
    const lineDecoder = new LineDecoder();
    const sseDecoder = new SSEDecoder();
    const rawLines: string[] = lineDecoder.decode(
      Buffer.from(rawChunksString, 'utf-8')
    );

    sseChunks = rawLines
      .map(chunk => sseDecoder.decode(chunk))
      .filter(sse => sse !== null)
      .filter(sse => !sse.data.startsWith('[DONE]'))
      .map(sse => JSON.parse(sse.data));

    async function* iterator(): AsyncGenerator<any> {
      for (const sseChunk of sseChunks) {
        yield sseChunk;
      }
    }
    originalChatCompletionStream = new OrchestrationStream(
      iterator,
      new AbortController()
    );
  });

  it('should wrap the raw chunk', async () => {
    let output = '';
    const asyncGenerator = OrchestrationStream._processChunk(
      originalChatCompletionStream
    );
    for await (const chunk of asyncGenerator) {
      expect(chunk).toBeDefined();
      output += chunk.getDeltaContent() ?? '';
    }
    expect(output).toMatchSnapshot();
  });

  it('should process the finish reasons', async () => {
    const logger = createLogger({
      package: 'orchestration',
      messageContext: 'stream-util'
    });
    const debugSpy = jest.spyOn(logger, 'debug');
    const asyncGeneratorChunk = OrchestrationStream._processChunk(
      originalChatCompletionStream
    );
    const asyncGeneratorFinishReason =
      OrchestrationStream._processOrchestrationStreamChunkResponse(
        new OrchestrationStream(
          () => asyncGeneratorChunk,
          new AbortController()
        ),
        new OrchestrationStreamResponse()
      );

    for await (const chunk of asyncGeneratorFinishReason) {
      expect(chunk).toBeDefined();
    }
    expect(debugSpy).toHaveBeenCalledWith('Choice 0: Stream finished.');
  });

  it('should process the token usage', async () => {
    const logger = createLogger({
      package: 'orchestration',
      messageContext: 'stream-util'
    });
    const debugSpy = jest.spyOn(logger, 'debug');
    const asyncGeneratorChunk = OrchestrationStream._processChunk(
      originalChatCompletionStream
    );
    const asyncGeneratorTokenUsage =
      OrchestrationStream._processOrchestrationStreamChunkResponse(
        new OrchestrationStream(
          () => asyncGeneratorChunk,
          new AbortController()
        ),
        new OrchestrationStreamResponse()
      );

    for await (const chunk of asyncGeneratorTokenUsage) {
      expect(chunk).toBeDefined();
    }
    expect(debugSpy).toHaveBeenCalledWith(
      expect.stringContaining('Token usage:')
    );
  });

  it('should transform the original stream to string stream', async () => {
    const asyncGeneratorChunk = OrchestrationStream._processChunk(
      originalChatCompletionStream
    );
    const chunkStream = new OrchestrationStream(
      () => asyncGeneratorChunk,
      new AbortController()
    );

    let output = '';
    for await (const chunk of chunkStream.toContentStream()) {
      expect(typeof chunk).toBe('string');
      output += chunk;
    }
    expect(output).toMatchSnapshot();
  });
});
