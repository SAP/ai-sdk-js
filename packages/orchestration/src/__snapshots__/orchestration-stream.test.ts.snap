// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orchestration chat completion stream should transform the original stream to string stream 1`] = `
"The SAP Cloud SDK is a comprehensive development toolkit designed to simplify and accelerate the creation of applications that integrate with SAP solutions, particularly those built on the SAP Business Technology Platform (BTP). It provides developers with libraries, tools, and best practices that streamline the process of connecting to SAP systems, such as S/4HANA and other services available on the SAP Cloud Platform.

Key features of the SAP Cloud SDK include:

1. **Simplified Connectivity**: The SDK offers pre-built libraries to easily interact with SAP services, providing capabilities for authentication, service consumption, and OData/REST client generation.

2. **Multi-cloud Support**: It supports multiple cloud environments, ensuring that applications remain flexible and can be deployed across various cloud providers.

3. **Best Practices and Guidelines**: The SDK includes best practices for development, ensuring high-quality, scalable, and maintainable code.

4. **Project Scaffolding and Code Samples**: Developers can quickly start their projects using provided templates and samples, accelerating the development process and reducing the learning curve.

5. **Extensive Documentation and Community Support**: Ample documentation, tutorials, and an active community help developers overcome challenges and adopt the SDK efficiently.

Overall, the SAP Cloud SDK is an essential tool for developers looking to build cloud-native applications and extensions that seamlessly integrate with SAP's enterprise solutions."
`;

exports[`Orchestration chat completion stream should wrap the raw chunk 1`] = `
"The SAP Cloud SDK is a comprehensive development toolkit designed to simplify and accelerate the creation of applications that integrate with SAP solutions, particularly those built on the SAP Business Technology Platform (BTP). It provides developers with libraries, tools, and best practices that streamline the process of connecting to SAP systems, such as S/4HANA and other services available on the SAP Cloud Platform.

Key features of the SAP Cloud SDK include:

1. **Simplified Connectivity**: The SDK offers pre-built libraries to easily interact with SAP services, providing capabilities for authentication, service consumption, and OData/REST client generation.

2. **Multi-cloud Support**: It supports multiple cloud environments, ensuring that applications remain flexible and can be deployed across various cloud providers.

3. **Best Practices and Guidelines**: The SDK includes best practices for development, ensuring high-quality, scalable, and maintainable code.

4. **Project Scaffolding and Code Samples**: Developers can quickly start their projects using provided templates and samples, accelerating the development process and reducing the learning curve.

5. **Extensive Documentation and Community Support**: Ample documentation, tutorials, and an active community help developers overcome challenges and adopt the SDK efficiently.

Overall, the SAP Cloud SDK is an essential tool for developers looking to build cloud-native applications and extensions that seamlessly integrate with SAP's enterprise solutions."
`;
